src/plugins/XHRUpload.js

```diff
@@ -248,19 +257,23 @@ module.exports = class XHRUpload extends Plugin {
   }
 
   uploadFiles (files) {
-    const promises = files.map((file, i) => {
+    const actions = files.map((file, i) => {
       const current = parseInt(i, 10) + 1
       const total = files.length
 
       if (file.error) {
-        return Promise.reject(new Error(file.error))
+        return () => Promise.reject(new Error(file.error))
       } else if (file.isRemote) {
-        return this.uploadRemote(file, current, total)
+        return this.uploadRemote.bind(this, file, current, total)
       } else {
-        return this.upload(file, current, total)
+        return this.upload.bind(this, file, current, total)
       }
     })
 
+    const promises = actions
+      .map(this.limitUploads)
+      .map((action) => action())
```

Comment:
alternative via @ifedapoolarewaju:

```js
const promises = actions.map((action) => {
  const limitedAction = this.limitUploads(action)
  return limitedAction()
})
```

Reply:
Yeah, to be honest, I find it easier to follow this last one, so if it’s all the same with you, could we stick with it? :)

Reply:
yeah i agree. pushed